Source/WebCore/platform/graphics/GraphicsContext.cpp

```diff
@@ -371,10 +364,7 @@ void GraphicsContext::drawImageBuffer(ImageBuffer& image, const FloatRect& desti
 void GraphicsContext::drawImageBuffer(ImageBuffer& image, const FloatRect& destination, const FloatRect& source, ImagePaintingOptions options)
 {
     InterpolationQualityMaintainer interpolationQualityForThisScope(*this, options.interpolationQuality());
-    FloatRect sourceScaled = source;
-    sourceScaled.scale(image.resolutionScale());
-    if (auto nativeImage = nativeImageForDrawing(image))
-        drawNativeImageInternal(*nativeImage, destination, sourceScaled, options);
+    image.draw(*this, destination, source, options);
 }
 
 void GraphicsContext::drawConsumingImageBuffer(RefPtr<ImageBuffer> image, const FloatPoint& destination, ImagePaintingOptions imagePaintingOptions)
```

Comment:
It seems weird that only some of the ImageBuffer drawing functions do this delegation? Why does drawConsumingImageBuffer, for instance, still use a NativeImage for this?

Reply:
You are right. `drawConsumingImageBuffer()` should not assume the underlying backend of the `ImageBuffer` is made of pixels. I will fix that.

Reply:
I am very interested to hear your thoughts on the point above about Image and ImageBuffer both now doing the same delegations.

Reply:
I think this can be done through these steps:

1. We need to make `drawImage()` returns `void`. And instead of returning `ImageDrawResult` the callers can just query the `decodingStatus` of the current frame.
2. We can combine the drawing of the `Image` and `ImageBuffer` functions in one set if we use `SourceImage` which is a variant of `Ref<Image>` and `Ref<ImageBuffer>`.
3. We can remove the `drawConsumingImageBuffer()` functions if make make the caller make  `context.drawImage(ImageBuffer::sinkIntoImage(WTFMove(imageBuffer)), ...)`.
4. We can have one function of `drawImage()` like this
```
void drawImage(SourceImage&, const FloatRect& destination, const std::optional<FloatRect>& source = std::nullopt, ImagePaintingOptions = { ImageOrientation::Orientation::FromImage });
```